application/controllers/FailedEmailController.php

```diff
@@ -0,0 +1,179 @@
+<?php
+
+class FailedEmailController extends LSBaseController
+{
+    /**
+     * this is part of renderWrappedTemplate implement in old responses.php
+     *
+     * @param string $view
+     * @return bool
+     */
+    public function beforeRender($view)
+    {
+        $surveyId = (int)App()->request->getParam('surveyid');
+        $this->aData['surveyid'] = $surveyId;
+        LimeExpressionManager::SetSurveyId($this->aData['surveyid']);
+        $this->layout = 'layout_questioneditor';
+
+        return parent::beforeRender($view);
+    }
+
+    /**
+     * @throws CHttpException
+     */
+    public function actionIndex(): void
+    {
+        $surveyId = sanitize_int(App()->request->getParam('surveyid'));
+        $permissions = [
+            'update' => Permission::model()->hasSurveyPermission($surveyId, 'responses', 'update'),
+            'delete' => Permission::model()->hasSurveyPermission($surveyId, 'responses', 'delete'),
+            'read'   => Permission::model()->hasSurveyPermission($surveyId, 'responses', 'read')
+        ];
+        if (!$surveyId) {
+            throw new CHttpException(403, gT("Invalid survey ID"));
+        }
+        if (!$permissions['read']) {
+            App()->user->setFlash('error', gT("You do not have permission to access this page."));
+            $this->redirect(['surveyAdministration/view', 'surveyid' => $surveyId]);
+        }
+
+        App()->getClientScript()->registerScriptFile('/application/views/failedEmail/javascript/failedEmail.js', LSYii_ClientScript::POS_BEGIN);
+        $failedEmailModel = FailedEmail::model();
+        $failedEmailModel->setAttributes(App()->getRequest()->getParam('FailedEmail'), false);
+        $pageSize = App()->request->getParam('pageSize') ?? App()->user->getState('pageSize', App()->params['defaultPageSize']);
+        $massiveAction = App()->getController()->renderPartial('/failedEmail/partials/massive_action_selector', [
+            'surveyId'    => $surveyId,
+            'permissions' => $permissions
+        ], true);
+
+
+        $this->render('failedEmail_index', [
+            'failedEmailModel' => $failedEmailModel,
+            'pageSize'         => $pageSize,
+            'massiveAction'    => $massiveAction,
+        ]);
+    }
+
+    /**
+     * @throws CHttpException
+     */
+    public function actionResend()
```

Comment:
61 lines long, you could probably break out some part to a separate function.

Reply:
dont have that locally